pymtl3/stdlib/rtl/queues.py

```diff
@@ -86,11 +86,298 @@ def construct( s, num_entries=2 ):
+
+    @s.update_on_edge
+    def up_reg():
+
+      if s.reset:
+        s.head  = PtrType(0)
+        s.tail  = PtrType(0)
+        s.count = CountType(0)
+
+      else:
+        s.head  = s.head_next if s.deq_xfer else s.head
+        s.tail  = s.tail_next if s.enq_xfer else s.tail
+        s.count = s.count + b1(1) if s.enq_xfer and not s.deq_xfer else \
+                  s.count - b1(1) if s.deq_xfer and not s.enq_xfer else \
+                  s.count
+
+#-------------------------------------------------------------------------
+# NormalQueueRTL
+#-------------------------------------------------------------------------
+
+class NormalQueueRTL( Component ):
+
+  def construct( s, EntryType, num_entries=2 ):
+
+    # Interface
+
+    s.enq   = EnqIfcRTL( EntryType )
+    s.deq   = DeqIfcRTL( EntryType )
+    s.count = OutPort( mk_bits( clog2( num_entries+1 ) ) )
+
+    # Components
+
+    s.ctrl  = NormalQueueCtrlRTL ( num_entries )
+    s.dpath = NormalQueueDpathRTL( EntryType, num_entries )
+
+    # Connect ctrl to data path
+
+    s.connect( s.ctrl.wen,     s.dpath.wen     )
+    s.connect( s.ctrl.waddr,   s.dpath.waddr   )
+    s.connect( s.ctrl.raddr,   s.dpath.raddr   )
+
+    # Connect to interface
+
+    s.connect( s.enq.en,  s.ctrl.enq_en   )
+    s.connect( s.enq.rdy, s.ctrl.enq_rdy  )
+    s.connect( s.deq.en,  s.ctrl.deq_en   )
+    s.connect( s.deq.rdy, s.ctrl.deq_rdy  )
+    s.connect( s.count,   s.ctrl.count    )
+    s.connect( s.enq.msg, s.dpath.enq_msg )
+    s.connect( s.deq.msg, s.dpath.deq_msg )
+
+  # Line trace
+
+  def line_trace( s ):
+    return "{}({}){}".format( s.enq, s.count, s.deq )
+
+#-------------------------------------------------------------------------
+# Ctrl for PipeQueue
+#-------------------------------------------------------------------------
+
+class PipeQueueCtrlRTL( Component ):
+
+  def construct( s, num_entries=2 ):
+
+    # Constants
+
+    s.num_entries = num_entries
+    s.last_idx    = num_entries - 1
+    addr_nbits    = clog2( num_entries   )
+    count_nbits   = clog2( num_entries+1 )
+    PtrType       = mk_bits( addr_nbits  )
+    CountType     = mk_bits( count_nbits )
+
+    # Interface
+
+    s.enq_en  = InPort ( Bits1     )
+    s.enq_rdy = OutPort( Bits1     )
+    s.deq_en  = InPort ( Bits1     )
+    s.deq_rdy = OutPort( Bits1     )
+    s.count   = OutPort( CountType )
+
+    s.wen     = OutPort( Bits1   )
+    s.waddr   = OutPort( PtrType )
+    s.raddr   = OutPort( PtrType )
+
+    # Registers
+
+    s.head = Wire( PtrType )
+    s.tail = Wire( PtrType )
+
+    # Wires
+
+    s.enq_xfer  = Wire( Bits1   )
+    s.deq_xfer  = Wire( Bits1   )
+    s.head_next = Wire( PtrType )
+    s.tail_next = Wire( PtrType )
+
+    # Connections
+
+    s.connect( s.wen,   s.enq_xfer )
+    s.connect( s.waddr, s.tail     )
+    s.connect( s.raddr, s.head     )
+
+    @s.update
+    def up_rdy_signals():
+      if ~s.reset:
+        s.deq_rdy = b1(1) if s.count > 0 else b1(0)
```

Comment:
Is it OK to use integer 0 here? @jsn1993

Reply:
I think it's ok for simulation but not sure if the translation pass enforce it ...

Reply:
uhhh it's kind of inconvenient to worry about types even for 0 and 1...

Reply:
I tried... need explicitly define the bitwidth for 0 here for translation (should be the same type as s.count)...

Reply:
Hmmm @ptpan what do you think?

Reply:
am I allowed to add bits with different width? like `s.head = s.head + b1(1)`?

Reply:
current translation requires exactly the same bitwidth on each side.